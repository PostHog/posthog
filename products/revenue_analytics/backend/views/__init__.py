from .revenue_analytics_base_view import RevenueAnalyticsBaseView
from .revenue_analytics_charge_view import RevenueAnalyticsChargeView
from .revenue_analytics_customer_view import RevenueAnalyticsCustomerView
from .revenue_analytics_invoice_item_view import RevenueAnalyticsInvoiceItemView
from .revenue_analytics_product_view import RevenueAnalyticsProductView

__all__ = [
    "RevenueAnalyticsBaseView",
    "RevenueAnalyticsChargeView",
    "RevenueAnalyticsCustomerView",
    "RevenueAnalyticsInvoiceItemView",
    "RevenueAnalyticsProductView",
]
