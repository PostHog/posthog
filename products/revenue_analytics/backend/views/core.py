from dataclasses import dataclass
from typing import Optional
from collections.abc import Callable, Iterable

from posthog.hogql import ast
from posthog.models.team.team import Team
from posthog.schema import DatabaseSchemaManagedViewTableKind
from posthog.warehouse.models.external_data_source import ExternalDataSource
import re


@dataclass
class SourceHandle:
    type: str  # 'events', 'stripe', 'chargebee', ...
    team: Team
    source: Optional[ExternalDataSource] = None


@dataclass
class BuiltQuery:
    # Stable key for naming: event name for events, table id (string) for warehouse
    key: str
    # Prefix used as source_label and to name the view
    prefix: str
    # HogQL AST for the view
    query: ast.Expr


# A builder is a function that takes a SourceHandle and returns an iterable of BuiltQuery objects
# This is the type of the builder functions in the sources/**/*.py files, transforming a source into a set of views.
# You can find all builder functions in the sources/**/*.py files, and they are registered in the sources/registry.py file.
Builder = dict[DatabaseSchemaManagedViewTableKind, Callable[[SourceHandle], Iterable[BuiltQuery]]]


def view_prefix_for_event(event: str) -> str:
    return f"revenue_analytics.events.{re.sub(r'[^a-zA-Z0-9]', '_', event)}"


def view_prefix_for_source(source: ExternalDataSource) -> str:
    if not source.prefix:
        return source.source_type.lower()
    prefix = source.prefix.strip("_")
    return f"{source.source_type.lower()}.{prefix}"


def view_name_for_event(event: str, suffix: str) -> str:
    return f"{view_prefix_for_event(event)}.{suffix}"


def view_name_for_source(source: ExternalDataSource, suffix: str) -> str:
    return f"{view_prefix_for_source(source)}.{suffix}"
