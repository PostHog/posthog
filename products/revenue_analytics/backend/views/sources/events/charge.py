from collections.abc import Iterable

from posthog.hogql import ast

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_event
from products.revenue_analytics.backend.views.sources.helpers import (
    events_expr_for_team,
    is_zero_decimal_in_stripe,
    currency_aware_divider,
    currency_aware_amount,
)
from posthog.hogql.database.schema.exchange_rate import (
    revenue_comparison_and_value_exprs_for_events,
    currency_expression_for_events,
)


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    team = handle.team

    if len(team.revenue_analytics_config.events) == 0:
        return

    generic_team_expr = events_expr_for_team(team)

    for event in team.revenue_analytics_config.events:
        prefix = view_prefix_for_event(event.eventName)

        comparison_expr, value_expr = revenue_comparison_and_value_exprs_for_events(
            team, event, do_currency_conversion=False
        )
        _, currency_aware_amount_expr = revenue_comparison_and_value_exprs_for_events(
            team,
            event,
            amount_expr=ast.Field(chain=["currency_aware_amount"]),
        )

        filter_exprs = [
            comparison_expr,
            generic_team_expr,
            ast.CompareOperation(
                op=ast.CompareOperationOp.NotEq,
                left=ast.Field(chain=["amount"]),
                right=ast.Constant(value=None),
            ),
        ]

        query = ast.SelectQuery(
            select=[
                ast.Alias(alias="id", expr=ast.Call(name="toString", args=[ast.Field(chain=["uuid"])])),
                ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
                ast.Alias(alias="timestamp", expr=ast.Field(chain=["timestamp"])),
                ast.Alias(alias="customer_id", expr=ast.Field(chain=["distinct_id"])),
                ast.Alias(
                    alias="invoice_id", expr=ast.Constant(value=None)
                ),  # Helpful for sources, not helpful for events
                ast.Alias(alias="session_id", expr=ast.Call(name="toString", args=[ast.Field(chain=["$session_id"])])),
                ast.Alias(alias="event_name", expr=ast.Field(chain=["event"])),
                ast.Alias(alias="original_currency", expr=currency_expression_for_events(team, event)),
                ast.Alias(alias="original_amount", expr=value_expr),
                # Being zero-decimal implies we will NOT divide the original amount by 100
                # We should only do that if we've tagged the event with `currencyAwareDecimal`
                # Otherwise, we'll just assume it's a non-zero-decimal currency
                ast.Alias(
                    alias="enable_currency_aware_divider",
                    expr=is_zero_decimal_in_stripe(ast.Field(chain=["original_currency"]))
                    if event.currencyAwareDecimal
                    else ast.Constant(value=True),
                ),
                currency_aware_divider(),
                currency_aware_amount(),
                ast.Alias(alias="currency", expr=ast.Constant(value=team.base_currency)),
                ast.Alias(alias="amount", expr=currency_aware_amount_expr),
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=ast.And(exprs=filter_exprs),
            order_by=[ast.OrderExpr(expr=ast.Field(chain=["timestamp"]), order="DESC")],
        )

        yield BuiltQuery(key=event.eventName, prefix=prefix, query=query)
