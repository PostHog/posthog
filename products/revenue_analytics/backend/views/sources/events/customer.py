from collections.abc import Iterable

from posthog.hogql import ast

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_event
from products.revenue_analytics.backend.views.sources.helpers import (
    events_expr_for_team,
    get_cohort_expr,
)


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    team = handle.team
    for event in team.revenue_analytics_config.events:
        prefix = view_prefix_for_event(event.eventName)

        events_query = ast.SelectQuery(
            distinct=True,
            select=[ast.Alias(alias="person_id", expr=ast.Field(chain=["events", "person", "id"]))],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=events_expr_for_team(team),
        )

        query = ast.SelectQuery(
            select=[
                ast.Alias(alias="id", expr=ast.Call(name="toString", args=[ast.Field(chain=["id"])])),
                ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
                ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
                ast.Alias(alias="name", expr=ast.Field(chain=["properties", "name"])),
                ast.Alias(alias="email", expr=ast.Field(chain=["properties", "email"])),
                ast.Alias(alias="phone", expr=ast.Field(chain=["properties", "phone"])),
                ast.Alias(alias="address", expr=ast.Field(chain=["properties", "address"])),
                ast.Alias(alias="metadata", expr=ast.Field(chain=["properties", "metadata"])),
                ast.Alias(alias="country", expr=ast.Field(chain=["properties", "$geoip_country_name"])),
                ast.Alias(alias="cohort", expr=get_cohort_expr("created_at")),
                ast.Alias(alias="initial_coupon", expr=ast.Constant(value=None)),
                ast.Alias(alias="initial_coupon_id", expr=ast.Constant(value=None)),
            ],
            select_from=ast.JoinExpr(
                table=ast.Field(chain=["persons"]),
                alias="persons",
                next_join=ast.JoinExpr(
                    table=events_query,
                    alias="events",
                    join_type="INNER JOIN",
                    constraint=ast.JoinConstraint(
                        constraint_type="ON",
                        expr=ast.CompareOperation(
                            left=ast.Field(chain=["id"]),
                            right=ast.Field(chain=["person_id"]),
                            op=ast.CompareOperationOp.Eq,
                        ),
                    ),
                ),
            ),
            order_by=[ast.OrderExpr(expr=ast.Field(chain=["created_at"]), order="DESC")],
        )

        yield BuiltQuery(key=event.eventName, prefix=prefix, query=query)
