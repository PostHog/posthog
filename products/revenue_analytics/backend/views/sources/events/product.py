from collections.abc import Iterable

from posthog.hogql import ast

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_event
from products.revenue_analytics.backend.views.sources.helpers import events_expr_for_team


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    team = handle.team
    for event in team.revenue_analytics_config.events:
        if not event.productProperty:
            continue

        prefix = view_prefix_for_event(event.eventName)

        events_query = ast.SelectQuery(
            distinct=True,
            select=[
                ast.Alias(alias="product_id", expr=ast.Field(chain=["events", "properties", event.productProperty]))
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=events_expr_for_team(team),
        )

        query = ast.SelectQuery(
            select=[
                ast.Alias(alias="id", expr=ast.Field(chain=["product_id"])),
                ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
                ast.Alias(alias="name", expr=ast.Field(chain=["product_id"])),
            ],
            select_from=ast.JoinExpr(table=events_query),
            order_by=[ast.OrderExpr(expr=ast.Field(chain=["id"]), order="ASC")],
        )

        yield BuiltQuery(key=event.eventName, prefix=prefix, query=query)
