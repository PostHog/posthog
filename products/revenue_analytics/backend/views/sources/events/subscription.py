from collections.abc import Iterable

from posthog.hogql import ast

from posthog.schema import SubscriptionDropoffMode
from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_event
from products.revenue_analytics.backend.views.sources.helpers import events_expr_for_team


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    team = handle.team
    for event in team.revenue_analytics_config.events:
        if event.subscriptionProperty is None:
            continue

        prefix = view_prefix_for_event(event.eventName)

        events_query = ast.SelectQuery(
            select=[
                ast.Alias(alias="person_id", expr=ast.Field(chain=["person", "id"])),
                ast.Alias(alias="subscription_id", expr=ast.Field(chain=["properties", event.subscriptionProperty])),
                ast.Alias(
                    alias="product_id",
                    expr=ast.Call(name="min", args=[ast.Field(chain=["properties", event.productProperty])])
                    if event.productProperty
                    else ast.Constant(value=None),
                ),
                ast.Alias(alias="min_timestamp", expr=ast.Call(name="min", args=[ast.Field(chain=["timestamp"])])),
                ast.Alias(alias="max_timestamp", expr=ast.Call(name="max", args=[ast.Field(chain=["timestamp"])])),
                ast.Alias(
                    alias="max_timestamp_plus_dropoff_days",
                    expr=ast.Call(
                        name="addDays",
                        args=[
                            ast.Field(chain=["max_timestamp"]),
                            ast.Constant(value=event.subscriptionDropoffDays),
                        ],
                    ),
                ),
            ],
            select_from=ast.JoinExpr(table=ast.Field(chain=["events"])),
            where=events_expr_for_team(team),
            group_by=[
                ast.Field(chain=["subscription_id"]),
                ast.Field(chain=["person_id"]),
            ],
        )

        query = ast.SelectQuery(
            select=[
                ast.Alias(alias="id", expr=ast.Field(chain=["subscription_id"])),
                ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
                ast.Alias(alias="plan_id", expr=ast.Constant(value=None)),
                ast.Alias(alias="product_id", expr=ast.Field(chain=["product_id"])),
                ast.Alias(alias="customer_id", expr=ast.Call(name="toString", args=[ast.Field(chain=["person_id"])])),
                ast.Alias(alias="status", expr=ast.Constant(value=None)),
                ast.Alias(alias="started_at", expr=ast.Field(chain=["min_timestamp"])),
                # If has an end date, but it's in the future, then just not include `ended_at`
                ast.Alias(
                    alias="ended_at",
                    expr=ast.Call(
                        name="if",
                        args=[
                            ast.CompareOperation(
                                op=ast.CompareOperationOp.Gt,
                                left=ast.Field(chain=["max_timestamp_plus_dropoff_days"]),
                                right=ast.Call(name="today", args=[]),
                            ),
                            ast.Constant(value=None),
                            ast.Field(chain=["max_timestamp"])
                            if event.subscriptionDropoffMode == SubscriptionDropoffMode.LAST_EVENT
                            else ast.Field(chain=["max_timestamp_plus_dropoff_days"]),
                        ],
                    ),
                ),
                ast.Alias(alias="metadata", expr=ast.Constant(value=None)),
            ],
            select_from=ast.JoinExpr(table=events_query),
            order_by=[ast.OrderExpr(expr=ast.Field(chain=["started_at"]), order="DESC")],
        )

        yield BuiltQuery(key=event.eventName, prefix=prefix, query=query)
