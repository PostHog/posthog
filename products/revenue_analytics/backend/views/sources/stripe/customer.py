from typing import cast
from collections.abc import Iterable

from posthog.hogql import ast

from posthog.temporal.data_imports.sources.stripe.constants import (
    CUSTOMER_RESOURCE_NAME as STRIPE_CUSTOMER_RESOURCE_NAME,
    INVOICE_RESOURCE_NAME as STRIPE_INVOICE_RESOURCE_NAME,
)
from posthog.warehouse.models.external_data_schema import ExternalDataSchema
from posthog.warehouse.models.table import DataWarehouseTable
from products.revenue_analytics.backend.views.sources.helpers import (
    extract_json_string,
    get_cohort_expr,
)

from products.revenue_analytics.backend.views.core import BuiltQuery, SourceHandle, view_prefix_for_source


def build(handle: SourceHandle) -> Iterable[BuiltQuery]:
    source = handle.source
    if source is None:
        return

    # Get all schemas for the source, avoid calling `filter` and do the filtering on Python-land
    # to avoid n+1 queries
    schemas = source.schemas.all()
    customer_schema = next((schema for schema in schemas if schema.name == STRIPE_CUSTOMER_RESOURCE_NAME), None)
    if customer_schema is None:
        return

    customer_schema = cast(ExternalDataSchema, customer_schema)
    if customer_schema.table is None:
        return

    invoice_schema = next((schema for schema in schemas if schema.name == STRIPE_INVOICE_RESOURCE_NAME), None)
    invoice_table = None
    if invoice_schema is not None:
        invoice_schema = cast(ExternalDataSchema, invoice_schema)
        invoice_table = invoice_schema.table
        if invoice_table is not None:
            invoice_table = cast(DataWarehouseTable, invoice_table)

    table = cast(DataWarehouseTable, customer_schema.table)
    prefix = view_prefix_for_source(source)

    query = ast.SelectQuery(
        select=[
            ast.Alias(alias="id", expr=ast.Field(chain=["outer", "id"])),
            ast.Alias(alias="source_label", expr=ast.Constant(value=prefix)),
            ast.Alias(alias="timestamp", expr=ast.Field(chain=["created_at"])),
            ast.Alias(alias="name", expr=ast.Field(chain=["name"])),
            ast.Alias(alias="email", expr=ast.Field(chain=["email"])),
            ast.Alias(alias="phone", expr=ast.Field(chain=["phone"])),
            ast.Alias(alias="address", expr=ast.Field(chain=["address"])),
            ast.Alias(alias="metadata", expr=ast.Field(chain=["metadata"])),
            ast.Alias(
                alias="country",
                expr=extract_json_string("address", "country"),
            ),
            ast.Alias(alias="cohort", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon", expr=ast.Constant(value=None)),
            ast.Alias(alias="initial_coupon_id", expr=ast.Constant(value=None)),
        ],
        select_from=ast.JoinExpr(
            alias="outer",
            table=ast.Field(chain=[table.name]),
        ),
    )

    # If there's an invoice table we can generate the cohort entry
    # by looking at the first invoice for each customer
    if invoice_table is not None:
        cohort_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "cohort"), None
        )
        if cohort_alias is not None:
            cohort_alias.expr = ast.Field(chain=["cohort"])

        initial_coupon_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon"),
            None,
        )
        if initial_coupon_alias is not None:
            initial_coupon_alias.expr = ast.Field(chain=["initial_coupon"])

        initial_coupon_id_alias: ast.Alias | None = next(
            (alias for alias in query.select if isinstance(alias, ast.Alias) and alias.alias == "initial_coupon_id"),
            None,
        )
        if initial_coupon_id_alias is not None:
            initial_coupon_id_alias.expr = ast.Field(chain=["initial_coupon_id"])

        if query.select_from is not None and (
            cohort_alias is not None or initial_coupon_alias is not None or initial_coupon_id_alias is not None
        ):
            query.select_from.next_join = ast.JoinExpr(
                alias="cohort_inner",
                table=ast.SelectQuery(
                    select=[
                        ast.Field(chain=["customer_id"]),
                        ast.Alias(alias="cohort", expr=get_cohort_expr("min(created_at)")),
                        ast.Alias(
                            alias="initial_coupon",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "name"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                        ast.Alias(
                            alias="initial_coupon_id",
                            expr=ast.Call(
                                name="argMin",
                                args=[
                                    extract_json_string("discount", "coupon", "id"),
                                    ast.Field(chain=["created_at"]),
                                ],
                            ),
                        ),
                    ],
                    select_from=ast.JoinExpr(alias="invoice", table=ast.Field(chain=[invoice_table.name])),
                    group_by=[ast.Field(chain=["customer_id"])],
                ),
                join_type="LEFT JOIN",
                constraint=ast.JoinConstraint(
                    constraint_type="ON",
                    expr=ast.CompareOperation(
                        left=ast.Field(chain=["cohort_inner", "customer_id"]),
                        right=ast.Field(chain=["outer", "id"]),
                        op=ast.CompareOperationOp.Eq,
                    ),
                ),
            )

    yield BuiltQuery(key=str(table.id), prefix=prefix, query=query)
