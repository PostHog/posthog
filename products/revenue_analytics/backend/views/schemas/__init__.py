from posthog.schema import DatabaseSchemaManagedViewTableKind
from ._definitions import Schema

from .subscription import SCHEMA as SUBSCRIPTION
from .charge import SCHEMA as CHARGE
from .customer import SCHEMA as CUSTOMER
from .product import SCHEMA as PRODUCT
from .revenue_item import SCHEMA as REVENUE_ITEM

SCHEMAS: dict[DatabaseSchemaManagedViewTableKind, Schema] = {
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_SUBSCRIPTION: SUBSCRIPTION,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_REVENUE_ITEM: REVENUE_ITEM,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CHARGE: CHARGE,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_CUSTOMER: CUSTOMER,
    DatabaseSchemaManagedViewTableKind.REVENUE_ANALYTICS_PRODUCT: PRODUCT,
}
