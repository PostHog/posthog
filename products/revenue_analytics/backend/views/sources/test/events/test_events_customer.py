from products.revenue_analytics.backend.views.core import SourceHandle
from products.revenue_analytics.backend.views.sources.events.customer import build
from products.revenue_analytics.backend.views.sources.test.events.base import EventsSourceBaseTest
from products.revenue_analytics.backend.views.schemas.customer import SCHEMA as CUSTOMER_SCHEMA


class TestCustomerEventsBuilder(EventsSourceBaseTest):
    def setUp(self):
        super().setUp()
        self.setup_revenue_analytics_events()

    def test_build_customer_queries_for_all_events(self):
        """Test building customer queries for all configured events."""
        handle = SourceHandle(type="events", team=self.team)

        queries = list(build(handle))

        # Should build one query per configured event
        self.assertEqual(len(queries), 2)

        # Test first query (purchase event)
        purchase_query = queries[0]
        self.assertBuiltQueryStructure(purchase_query, "purchase", "revenue_analytics.events.purchase")

        # Print and snapshot the generated HogQL AST query
        query_sql = purchase_query.query.to_hogql()
        self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_customer_query_uses_distinct_persons(self):
        """Test that customer queries use DISTINCT to get unique persons."""
        handle = SourceHandle(type="events", team=self.team)

        queries = list(build(handle))

        # Test subscription_charge query
        subscription_query = queries[1]
        self.assertBuiltQueryStructure(
            subscription_query, "subscription_charge", "revenue_analytics.events.subscription_charge"
        )

        # Should use DISTINCT in the events subquery to get unique persons
        query_sql = subscription_query.query.to_hogql()
        self.assertIn("SELECT DISTINCT", query_sql)
        self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)

    def test_build_with_no_events_configured(self):
        """Test that build returns empty list when no events are configured."""
        # Clear revenue analytics events
        self.clear_events()

        handle = SourceHandle(type="events", team=self.team)
        queries = list(build(handle))

        self.assertEqual(len(queries), 0)

    def test_query_structure_contains_required_fields(self):
        """Test that the generated query contains all required fields."""
        handle = SourceHandle(type="events", team=self.team)

        queries = list(build(handle))
        purchase_query = queries[0]

        self.assertQueryContainsFields(purchase_query.query, CUSTOMER_SCHEMA)

    def test_customer_query_structure(self):
        """Test that customer queries have the expected structure."""
        handle = SourceHandle(type="events", team=self.team)

        queries = list(build(handle))

        # Customer queries should join persons with events
        purchase_query = queries[0]
        query_sql = purchase_query.query.to_hogql()
        self.assertQueryMatchesSnapshot(query_sql, replace_all_numbers=True)
