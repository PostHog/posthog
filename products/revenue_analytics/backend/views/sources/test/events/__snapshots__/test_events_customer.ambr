# serializer version: 1
# name: TestCustomerEventsBuilder.test_build_customer_queries_for_all_events
  '''
  SELECT toString(id) AS id,
         'revenue_analytics.events.purchase' AS source_label,
         created_at AS timestamp,
         properties.name AS name,
         properties.email AS email,
         properties.phone AS phone,
         properties.address AS address,
         properties.metadata AS metadata,
         properties.$geoip_country_name AS country,
         formatDateTime(toStartOfMonth(created_at), '%Y-%m') AS cohort,
         NULL AS initial_coupon,
         NULL AS initial_coupon_id
  FROM persons AS persons
  INNER JOIN
    (SELECT DISTINCT events.person.id AS person_id
     FROM events
     WHERE true) AS events ON equals(id, person_id)
  ORDER BY created_at DESC
  '''
# ---
# name: TestCustomerEventsBuilder.test_customer_query_structure
  '''
  SELECT toString(id) AS id,
         'revenue_analytics.events.purchase' AS source_label,
         created_at AS timestamp,
         properties.name AS name,
         properties.email AS email,
         properties.phone AS phone,
         properties.address AS address,
         properties.metadata AS metadata,
         properties.$geoip_country_name AS country,
         formatDateTime(toStartOfMonth(created_at), '%Y-%m') AS cohort,
         NULL AS initial_coupon,
         NULL AS initial_coupon_id
  FROM persons AS persons
  INNER JOIN
    (SELECT DISTINCT events.person.id AS person_id
     FROM events
     WHERE true) AS events ON equals(id, person_id)
  ORDER BY created_at DESC
  '''
# ---
# name: TestCustomerEventsBuilder.test_customer_query_uses_distinct_persons
  '''
  SELECT toString(id) AS id,
         'revenue_analytics.events.subscription_charge' AS source_label,
         created_at AS timestamp,
         properties.name AS name,
         properties.email AS email,
         properties.phone AS phone,
         properties.address AS address,
         properties.metadata AS metadata,
         properties.$geoip_country_name AS country,
         formatDateTime(toStartOfMonth(created_at), '%Y-%m') AS cohort,
         NULL AS initial_coupon,
         NULL AS initial_coupon_id
  FROM persons AS persons
  INNER JOIN
    (SELECT DISTINCT events.person.id AS person_id
     FROM events
     WHERE true) AS events ON equals(id, person_id)
  ORDER BY created_at DESC
  '''
# ---
